### PR TITLE
Tag the differentiable variables in RNN operator

### PR DESCRIPTION
**Pull request by rajeevnalawadi**
Adding tags to explicitly define the differentiable variables for RNN. 
This PR uses the differentiable tag and defines the input & output variables in LSTM, GRU to support training

### CODE REVIEW

**Comment by wschin** [comment]
>             OpSchema::Optional)

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:239 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>             OpSchema::Optional)

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:368 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>             OpSchema::Optional)

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:503 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>             OpSchema::Optional)

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:519 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>             OpSchema::Optional));

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:527 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>         OpSchema::Optional);

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:122 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>         OpSchema::Optional);

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:115 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>         OpSchema::Optional);

Non-differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:101 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by wschin** [comment]
>         "T");

Differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:93 in 772db2c. [](commit_id = 772db2c41b526b7c78f37abe69a137983673c326, deletion_comment = False)

**Comment by rajeevnalawadi** [comment]
> "T");
> 
> 
> Differentiable?
> 
> Refers to: onnx/defs/rnn/defs.cc:93 in 772db2c. [](commit_id = 772db2c, deletion_comment = False)

**Comment by rajeevnalawadi** [comment]
Thanks Wei Sheng for catching some of the missing differentiable tags, have updated

File: onnx/defs/rnn/defs.cc [hunk 1]
```diff
@@ -90,15 +90,22 @@ std::function<void(OpSchema&)> RNNDocGenerator(const char* /*name*/) {
...
+        OpSchema::Differentiable);
     schema.Input(
         4,
         "sequence_lens",
         "Optional tensor specifying lengths of the sequences in a batch. "
         "If not specified - assumed all sequences in the batch to have "
         "length `seq_length`. It has shape `[batch_size]`.",
         "T1",
-        OpSchema::Optional);
+        OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:105]:
```suggestion
        OpSchema::Optional,
```
This should be `OpSchema::Optional`. All `optional` variables need to remain `optional`.

File: onnx/defs/rnn/defs.cc [hunk 2]
```diff
@@ -112,14 +119,20 @@ std::function<void(OpSchema&)> RNNDocGenerator(const char* /*name*/) {
         "A tensor that concats all the intermediate output values of the hidden. "
         "It has shape `[seq_length, num_directions, batch_size, hidden_size]`. ",
         "T",
-        OpSchema::Optional);
+        OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:122]:
```suggestion
        OpSchema::Optional,
```

File: onnx/defs/rnn/defs.cc [hunk 3]
```diff
@@ -112,14 +119,20 @@ std::function<void(OpSchema&)> RNNDocGenerator(const char* /*name*/) {
...
+        1,
+        OpSchema::Differentiable);
     schema.Output(
         1,
         "Y_h",
         "The last output value of the hidden. It has shape "
         "`[num_directions, batch_size, hidden_size]`.",
         "T",
-        OpSchema::Optional);
+        OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:132]:
```suggestion
        OpSchema::Optional,
```

File: onnx/defs/rnn/defs.cc [hunk 4]
```diff
@@ -228,7 +249,10 @@ ONNX_OPERATOR_SET_SCHEMA(
             "`[num_directions, 2*hidden_size]`. Optional: If not specified - assumed "
             "to be 0.",
             "T",
-            OpSchema::Optional)
+            OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:252]:
```suggestion
            OpSchema::Optional,
```

File: onnx/defs/rnn/defs.cc [hunk 5]
```diff
@@ -349,7 +381,10 @@ ONNX_OPERATOR_SET_SCHEMA(
             "has shape `[num_directions, 6*hidden_size]`. Optional: If not specified "
             "- assumed to be 0",
             "T",
-            OpSchema::Optional)
+            OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:384]:
```suggestion
            OpSchema::Optional,
```

File: onnx/defs/rnn/defs.cc [hunk 6]
```diff
@@ -476,7 +519,10 @@ ONNX_OPERATOR_SET_SCHEMA(
             "tensor has shape `[num_directions, 8*hidden_size]`. Optional: If not "
             "specified - assumed to be 0.",
             "T",
-            OpSchema::Optional)
+            OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:522]:
```suggestion
            OpSchema::Optional,
```

File: onnx/defs/rnn/defs.cc [hunk 7]
```diff
@@ -492,13 +538,19 @@ ONNX_OPERATOR_SET_SCHEMA(
             "`[num_directions, 3*hidde_size]`. Optional: If not specified - "
             "assumed to be 0.",
             "T",
-            OpSchema::Optional)
+            OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:541]:
```suggestion
            OpSchema::Optional,
```

File: onnx/defs/rnn/defs.cc [hunk 8]
```diff
@@ -492,13 +538,19 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            OpSchema::Differentiable)
         .FillUsing(RNNDocGenerator("LSTM"))
         .Output(
             2,
             "Y_c",
             "The last output value of the cell. It has shape "
             "`[num_directions, batch_size, hidden_size]`.",
             "T",
-            OpSchema::Optional));
+            OpSchema::Single,
```
Comment (wschin) [inline onnx/defs/rnn/defs.cc:552]:
```suggestion
            OpSchema::Optional,
```

File: docs/Changelog.md [hunk 9]
```diff
@@ -7285,29 +7285,29 @@ This version of the operator has been available since version 7 of the default O
 <dd>Number of neurons in the hidden layer</dd>
 </dl>
 
-#### Inputs (3 - 6)
+#### Inputs (5 - 6)
```
Comment (wschin) [inline docs/Changelog.md:7288]: This should not be changed. The two differentiable inputs should also be `Optional`.

File: docs/Changelog.md [hunk 10]
```diff
@@ -6708,29 +6708,29 @@ This version of the operator has been available since version 7 of the default O
...
 <dd>The bias tensor for the gates. Concatenation of `[Wb[zrh], Rb[zrh]]` and `[WBb[zrh], RBb[zrh]]` (if bidirectional) along dimension 0. This tensor has shape `[num_directions, 6*hidden_size]`. Optional: If not specified - assumed to be 0</dd>
-<dt><tt>sequence_lens</tt> (optional) : T1</dt>
+<dt><tt>sequence_lens</tt> (non-differentiable) : T1</dt>
 <dd>Optional tensor specifying lengths of the sequences in a batch. If not specified - assumed all sequences in the batch to have length `seq_length`. It has shape `[batch_size]`.</dd>
 <dt><tt>initial_h</tt> (optional) : T</dt>
 <dd>Optional initial value of the hidden. If not specified - assumed to be 0. It has shape `[num_directions, batch_size, hidden_size]`.</dd>
 </dl>
 
-#### Outputs (0 - 2)
+#### Outputs
```
Comment (wschin) [inline docs/Changelog.md:6728]: This should not be changed. Please double-check if the two outputs are still `Optional` (instead of `Single`) in defs.cc.

Comment (rajeevnalawadi) [reply docs/Changelog.md]: With the required differentiable variables highlighted as optional based on initial definition, the errors have been resolved

**Comment by rajeevnalawadi** [comment]
RNN Example for verifying the results based on Pytorch framework

import torch
from torch.autograd import Variable
import numpy as np
import pylab as pl
import torch.nn.init as init

dtype = torch.FloatTensor
input_size, hidden_size, output_size = 7, 6, 1
epochs = 300
seq_length = 20
lr = 0.1
data_time_steps = np.linspace(2, 10, seq_length + 1)
data = np.sin(data_time_steps)
data.resize((seq_length + 1, 1))

x = Variable(torch.Tensor(data[:-1]).type(dtype), requires_grad=False)
y = Variable(torch.Tensor(data[1:]).type(dtype), requires_grad=False)

# We will generate training data, where x is the input data sequence and y is required target sequence.

w1 = torch.FloatTensor(input_size, 
hidden_size).type(dtype)
init.normal(w1, 0.0, 0.4)
w1 = Variable(w1, requires_grad = True)
w2 = torch.FloatTensor(hidden_size, output_size).type(dtype)
init.normal(w2, 0.0, 0.3)
w2 = Variable(w2, requires_grad = True)

# create a function for feed forward
def forward(input, context_state, w1, w2):
   xh = torch.cat((input, context_state), 1)
   context_state = torch.tanh(xh.mm(w1))
   out = context_state.mm(w2)
   return (out, context_state)

for i in range(epochs):
   total_loss = 0
   context_state = Variable(torch.zeros((1, hidden_size)).type(dtype), requires_grad = True)
   for j in range(x.size(0)):
      input = x[j:(j+1)]
      target = y[j:(j+1)]
      (pred, context_state) = forward(input, context_state, w1, w2)
      loss = (pred - target).pow(2).sum()/2
      total_loss += loss
      loss.backward()
      w1.data -= lr * w1.grad.data
      w2.data -= lr * w2.grad.data
      w1.grad.data.zero_()
      w2.grad.data.zero_()
      context_state = Variable(context_state.data)
   if i % 10 == 0:
      print("Epoch: {} loss {}".format(i, total_loss.data[0]))

context_state = Variable(torch.zeros((1, hidden_size)).type(dtype), requires_grad = False)
predictions = []

for i in range(x.size(0)):
   input = x[i:i+1]
   (pred, context_state) = forward(input, context_state, w1, w2)
   context_state = context_state
   predictions.append(pred.data.numpy().ravel()[0])

# Plot the sine wave
pl.scatter(data_time_steps[:-1], x.data.numpy(), s = 90, label = "Actual")
pl.scatter(data_time_steps[1:], predictions, label = "Predicted")
pl.legend()
pl.show()

**Comment by wschin** [comment]
>         OpSchema::Optional);

non-differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:115 in d150b82. [](commit_id = d150b82e3a822357243c8b96c3a95edd593ed875, deletion_comment = False)

**Comment by wschin** [comment]
>             OpSchema::Optional)

non-differentiable?

---
Refers to: onnx/defs/rnn/defs.cc:532 in d150b82. [](commit_id = d150b82e3a822357243c8b96c3a95edd593ed875, deletion_comment = False)
